src/shims/env.rs

```diff
@@ -104,11 +103,52 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
         if let Some(old) = success {
             if let Some(var) = old {
-                this.memory_mut().deallocate(var, None, MiriMemoryKind::Env.into())?;
+                this.memory_mut()
+                    .deallocate(var, None, MiriMemoryKind::Env.into())?;
             }
             Ok(0)
         } else {
             Ok(-1)
         }
     }
+
+    fn getcwd(
+        &mut self,
+        buf_op: OpTy<'tcx, Tag>,
+        size_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, Scalar<Tag>> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`getcwd` not available when isolation is enabled")
+        }
+
+        let tcx = &{ this.tcx.tcx };
+
+        let buf = this.force_ptr(this.read_scalar(buf_op)?.not_undef()?)?;
+        let size = this.read_scalar(size_op)?.to_usize(&*this.tcx)?;
+        // If we cannot get the current directory, we return null
+        match env::current_dir() {
+            Ok(cwd) => {
+                // It is not clear what happens with non-utf8 paths here
+                let mut bytes = cwd.display().to_string().into_bytes();
+                // If the buffer is smaller or equal than the path, we return null.
+                if (bytes.len() as u64) < size {
+                    // We add a `/0` terminator
+                    bytes.push(0);
```

Comment:
This makes the buffer one byte longer. Doesn't that break the `size` logic?

Reply:
I think it doesn't because `bytes.len()` is strictly smaller than `size`. So when you add the null terminator, `bytes.len()` would be smaller or equal to `size` so it still fits inside the buffer.

Reply:
Ah makes sense. But this needs a comment in the code explaining that argument.

Reply:
https://github.com/rust-lang/miri/pull/983/commits/cb51e7dc6efbe15745e85bfe04aaec819dcea11e